### ISSUE TITLE
Slow AxisVisual performance

### ISSUE THREAD

**Issue by Tillsten** [vispy/vispy#1609]
Hello, 
when running the simple toy-example below, showing an updating lineplot with two axis, on my computer, I only get around 30fps. Without linking the axis to the view, the speed is back to normal 60 fps. 

Vispy-Info
```
0.6.0.dev0
Platform: Windows-10-10.0.17763-SP0
Python:   3.6.5 |Anaconda, Inc.| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)]
NumPy:    1.16.2
Backend:  PyQt5
pyqt4:    None
pyqt5:    ('PyQt5', '5.9.2', '5.9.4')
pyside:   None
pyside2:  None
pyglet:   None
glfw:     glfw (3, 1, 0)
sdl2:     None
wx:       None
egl:      None
osmesa:   None
_test:    None
```

The example

```python
import sys
import numpy as np
import vispy
from vispy import scene, app
print(vispy.__version__)
print(vispy.sys_info())

canvas = scene.SceneCanvas(keys='interactive', size=(2*600, 2*600), show=True)
canvas.measure_fps()

grid = canvas.central_widget.add_grid(margin=10)
grid.spacing = 0
title = scene.Label("Plot Title", color='white')
title.height_max = 40
grid.add_widget(title, row=0, col=0, col_span=2)

yaxis = scene.AxisWidget(orientation='left',
                         axis_label='Y Axis',
                         axis_font_size=20,
                         axis_label_margin=50,
                         tick_label_margin=5)
yaxis.width_max = 80
grid.add_widget(yaxis, row=1, col=0)

xaxis = scene.AxisWidget(orientation='bottom',
                         axis_label='X Axis',
                         axis_font_size=20,
                         axis_label_margin=50,
                         tick_label_margin=5)

xaxis.height_max = 80
N = 500
grid.add_widget(xaxis, row=2, col=1)
right_padding = grid.add_widget(row=1, col=2, row_span=1)
right_padding.width_max = 50

view = grid.add_view(row=1, col=1, border_color='white')

data = np.zeros((N, 2))
plot = scene.Line(data, parent=view.scene)

view.camera = 'panzoom'
plot.set_data(data, width=4, color='g')

xaxis.link_view(view)
yaxis.link_view(view)
import time
t0 = time.time()

def update(up):
    #print(up.count)
    global data
    data[:, 0] = np.linspace(0, 10, N)+time.time()-t0
    data[:, 1] = np.sin(data[:, 0])+0.2*np.random.normal(size=(N))
    plot.set_data(data)
    view.camera.set_range(y=(data[:, 1].max(), data[:, 1].min()), x=(data[:, 0].max(), data[:, 0].min()))


t = app.Timer()
t.connect(update)
t.start(iterations=60*15)
t.events.stop.connect(lambda x: app.quit())

if __name__ == '__main__' and sys.flags.interactive == 0:
    app.run()
```

**Comment by Tillsten** [comment]
Profiling via cProfiler gives:
![grafik](https://user-images.githubusercontent.com/189880/54872110-6465db80-4dbf-11e9-81c3-aba27ca7a6c9.png)

**Comment by Tillsten** [comment]
Hardware is a i5-3570K CPU with a 1070ti, which should be powerful enough. Changing the backend and or changing the _text visual method to 'gpu' do not seem to affect the fps.

**Comment by kmuehlbauer** [comment]
@Tillsten It would be really helpful, if you can add some words on how you installed vispy (including dependencies, compilers). It would also help, if you could re-create this and add all possible outputs/logs of the installation process.
As mentioned above I have the same behaviour on my dual boot Ryzen Box. But kubuntu provides normal behaviour, whereas Win 10 does not. So it's not a hardware problem nor a general vispy problem. It seems a OS or dependency problem.

**Comment by Tillsten** [comment]
I tried both fresh packages from conda-base and conda-forge. Vispy was installed by pip, there is not much freedom when it comes to the compiler, one has to use the win10 sdk compiler.

**Comment by kmuehlbauer** [comment]
@Tillsten Thanks. I'm currently trying to avoid conda/anaconda by installing python from python.org. But vispy dependencies are hard to get via pip. Anyway still trying. I checked and reinstalled my windows radeon drivers, all good. OpenGL test programs working as expected. 

Do you have other (WIN) machines at your disposal where you can try this?

**Comment by codypiersall** [comment]
I'm getting ~58 FPS on Win10 on vispy master.

**Comment by kmuehlbauer** [comment]
@codypiersall Can we assume, that FPS values below 60 indicate that the application has some performance problem? I mean on Linux and Mac we get stable 60 FPS with this test case. Also the VM win 10 has 60 FPS. Only WIN 10 has degradations.

**Comment by djhoese** [comment]
FYI I was getting steady 60 FPS on my mac as long as I wasn't doing anything else. Every time I switched windows, opened a new chrome tab, or really did anything it would dip to ~55FPS for a moment.

**Comment by codypiersall** [comment]
My guess is that for this test, 58 is close enough to 60 that I would say it's working okay on my computer.  My profiling results were consistent with @Tillsten's though, where most time was spent in [`_get_ticks_talbot`](https://github.com/vispy/vispy/blob/d96c26ec3093b9ea576bed95072a9c3ef6be1e27/vispy/visuals/axis.py#L500-L577).

**Comment by codypiersall** [comment]
On my home computer I'm getting 60 fps on Linux (Ubuntu 16.04), but about 45 fps on Windows.

**Comment by kmuehlbauer** [comment]
The question is, why do we get reasonable values for linux/mac but not for windows?

**Comment by Tillsten** [comment]
Does the profiling result differ between windows and linux?

**Comment by neurochen** [comment]
I noticed this issue several months ago. As axis updating is not frequent in my case, so the issue is not reported immediately. Here's my test results today on Windows10 and GTX 1080 Ti. It's around 50 Hz with linked axes, but ran back to 60 hz once disabling the axis update.
'''python
Platform: Windows-10-10.0.14393-SP0
Python:   3.7.2 (default, Jan  2 2019, 17:07:39) [MSC v.1915 64 bit (AMD64)]
NumPy:    1.15.4
Backend:  PyQt5
pyqt4:    None
pyqt5:    ('PyQt5', '5.11.3', '5.11.2')
pyside:   None
pyside2:  None
pyglet:   pyglet 1.3.2
glfw:     glfw (3, 2, 1)
sdl2:     None
wx:       wxPython 4.0.4
egl:      None
osmesa:   None
_test:    None

1.2 FPS
47.1 FPS
51.0 FPS
50.2 FPS
48.7 FPS
50.0 FPS
50.6 FPS
50.4 FPS
49.8 FPS
'''

**Comment by kmuehlbauer** [comment]
@neurochen Thanks for looking into this. Could you explain how you install vispy and dependencies in your machine? Only to make the installation process easier for me.

But nevertheless your results agree with the other results so far. What I still do not understand where this discrepancy between OS results from.

**Comment by neurochen** [comment]
This cprofile output might help. It seems related to the vertex buffer object for axis texts. 

![image](https://user-images.githubusercontent.com/10821596/55075362-2be92a80-5069-11e9-9169-4747d9edccec.png)

**Comment by neurochen** [comment]
> @neurochen Thanks for looking into this. Could you explain how you install vispy and dependencies in your machine? Only to make the installation process easier for me.
> 
> But nevertheless your results agree with the other results so far. What I still do not understand where this discrepancy between OS results from.

Regarding your question, my installation is as follows: 
- miniconda
- conda env with python=3.7, numpy, pyqt
- pipinstall git+https://github.com/vispy/vispy.git

**Comment by kmuehlbauer** [comment]
> Regarding your question, my installation is as follows:

OK, so I assume you have some Visual C build tools installed to get the cython extension compiled (text_sdf)? This is exactly the same setup as mine.

**Comment by neurochen** [comment]
> > Regarding your question, my installation is as follows:
> 
> OK, so I assume you have some Visual C build tools installed to get the cython extension compiled (text_sdf)? This is exactly the same setup as mine.

Yes, I have a visual studio community 2017 installed, which comes with a build tool as well. Just went through the _text_to_vbo function, without noticing any code related to performance drop actually.

**Comment by kmuehlbauer** [comment]
So for profiling, just for me and @djhoese (kudos to @Tillsten):

Invoking external Profiler:
`python -m cProfile -o test.perf test.py`

using `snakeviz`:

`conda install snakeviz`
`snakeviz test.perf`

There should be some internal profiling possible, but I currently lack this knowledge.

**Comment by kmuehlbauer** [comment]
Here is 20 seconds profile on my linux machine
[vispy_axis_slow.perf.zip](https://github.com/vispy/vispy/files/3013584/vispy_axis_slow.perf.zip)

**Comment by kmuehlbauer** [comment]
@Tillsten Can you check with my linux profiling from above, where the changes are, compared to your profiling. From the first glance, it seems that too much time is spent in `_text_to_vbo` and `_get_ticks_talbot` (at least more than on my system).

But also the `built-in method swapBuffers` consumes much more time on your system than mine. Also some other differences are there. This will be hard nut to crack. 

What I found so far, which might be related somehow: We have some Profiler initializations within vispy (Canvas, LineVisual IIRC). Not sure how to debug this.
